### PR TITLE
fix: makes Logger.With immutable.

### PR DESCRIPTION
**Pull request by jcchavezs**
Currently Logger.With isn't immutable leading us to mutable loggers. For example, on every request we want to create a new logger with the request ID but with current logic it will mutate the whole logger instead of creating a new one.

Related to https://github.com/openfga/openfga-envoy/pull/17

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Review by adriantam — COMMENTED** [review]
Please update the CHANGELOG.md unreleased section to reflect this change.

File: pkg/logger/logger.go [hunk 1]
```diff
@@ -43,8 +44,8 @@ type ZapLogger struct {
 
 var _ Logger = (*ZapLogger)(nil)
 
-func (l *ZapLogger) With(fields ...zap.Field) {
-	l.Logger = l.Logger.With(fields...)
+func (l *ZapLogger) With(fields ...zap.Field) Logger {
```
Comment (miparnisari) [inline pkg/logger/logger.go:47]: I can't find any usages of this function that is not unit tests. What is the value of this change?

Comment (jcchavezs) [reply pkg/logger/logger.go:47]: The value of the change is that it fixes the behavior. The potential value of this change is that we can embed per request data into logs like I am aiming in https://github.com/openfga/openfga-envoy/pull/17

Comment (jcchavezs) [reply pkg/logger/logger.go:47]: BTW it is being used here https://github.com/openfga/openfga/blob/main/pkg/logger/logger_test.go#L111

Comment (adriantam) [reply pkg/logger/logger.go:50]: Check that it matches with the original zap logger interface - https://github.com/uber-go/zap/blob/d6b2d14fc541ed17b31b794afc139f03648842ca/logger.go#L180

**Review by adriantam — APPROVED** [review]

File: pkg/logger/logger.go (hunk 1 above)
Comment (miparnisari) [reply pkg/logger/logger.go:50]: OK, understood now, the goal is to be able to do this:

```go
reqID := req.Attributes.GetRequest().GetHttp().GetHeaders()["x-request-id"]
logger := logger.With("request_id", reqID)
// ...
if err != nil {
		logger.Error("some error")
}
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -178,23 +183,22 @@ This consistency data should be used to help determine whether any form of cachi
...
 
-
 ## [1.5.3] - 2024-04-16
 
 [Full changelog](https://github.com/openfga/openfga/compare/v1.5.2...v1.5.3)
 
 ### Added
 
-* Apply tags to requests that have been intentionally throttled (https://github.com/openfga/openfga/pull/1531). This will add a new log field titled "throttled" to such requests.
+* Apply tags to requests that have been intentionally throttled (<https://github.com/openfga/openfga/pull/1531>). This will add a new log field titled "throttled" to such requests.
```
Comment (miparnisari) [inline CHANGELOG.md:192]: 🤔

**Review by miparnisari — DISMISSED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (jcchavezs) [reply CHANGELOG.md]: undoes all format changes.

**Comment by jcchavezs** [comment]
I reverted all format changes in the CHANGELOG.md and now I need new approvals cc @miparnisari @adriantam

**Review by miparnisari — APPROVED** [review]

**Review by adriantam — APPROVED** [review]

**Pull request merged**